KAFKA-17304; RaftClient API makes writing to log explicit

jsancio opened the pull request:
RaftClient API is changed to separate the batch accumulation (`RaftClient#prepareAppend`) from scheduling the append of accumulated batches (`RaftClient#schedulePrepatedAppend`) to the KRaft log. This change is needed to better match the controller's flow of replaying the generated records before replicating them. When the controller replay records it needs to know the offset associated with the record. To compute a table offset the KafkaClient needs to be aware of the records and their log position.

The controller uses this new API by generated the cluster metadata records, compute their offset using RaftClient#prepareAppend, replay the records in the state machine, and finally allowing KRaft to append the records with RaftClient#schedulePreparedAppend.

To implement this API the BatchAccumulator is changed to also support this access pattern. This is done by adding a drainOffset to the implementation. The batch accumulator is allowed to return any record and batch that is less than the drain offset.

Lastly, this change also removes some functionality that is no longer needed like non-atomic appends and validation of the base offset.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jsancio committed:
KAFKA-17304; RaftClient API makes writing to log explicit

cmccabe commented on metadata/src/main/java/org/apache/kafka/controller/errors/EventHandlerExceptionInfo.java:
```diff
@@ -88,11 +86,6 @@ public static EventHandlerExceptionInfo fromInternal(
             return new EventHandlerExceptionInfo(false, false, internal,
                 new PolicyViolationException("Unable to perform excessively large batch " +
                     "operation."));
-        } else if (internal instanceof UnexpectedBaseOffsetException) {
```
So this will be treated as an UnknownServerException if it does happen? I guess that's reasonable

jsancio replied:
I actually removed this exception. KRaft is not checking the base offset anymore so it doesn't throw this exception anymore.

cmccabe commented on metadata/src/test/java/org/apache/kafka/metalog/LocalLogManager.java:
```diff
@@ -729,52 +714,23 @@ public synchronized OptionalLong highWatermark() {
+            throw new BufferAllocationException("Test asked to fail the next scheduleAppend");
         }
+
+        return shared.tryAppend(nodeId, leader.epoch(), batch);
```
This isn't correct, since it means we are really appending the records immediately rather than waiting until `scheduleFlush`. You should instead just put the records into a temporary area and flush them in `scheduleFlush`.

jsancio replied:
Yes. It is not correct but it is not trivial to implement. I tried to implement it but ran into many issues. I filed https://issues.apache.org/jira/browse/KAFKA-17339 to revisit this.

cmccabe commented on metadata/src/test/java/org/apache/kafka/metalog/LocalLogManager.java:
```diff
@@ -729,52 +714,23 @@ public synchronized OptionalLong highWatermark() {
-        }
-        return shared.tryAppend(nodeId, leader.epoch(), requiredEndOffset, batch);
-    }
+    public void scheduleFlush() { }
```
Not correct, the records should be getting added to "shared" here

jsancio replied:
See my other comment. I filed https://issues.apache.org/jira/browse/KAFKA-17339.

cmccabe reviewed (approved):
The overall idea looks good to me, as does the API.

However,

- We need a test where we schedule an append but then don't ever call flush (perhaps combined with a failover to a new leader)

- LocalLogManager implementation is not correct (it should be easy to do this right by creating a list inside scheduleAppend and using it in scheduleFlush)

- JavaDoc TODO :)

mumrah commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -3302,19 +3297,34 @@ private long append(int epoch, List<T> records, OptionalLong requiredBaseOffset,
         BatchAccumulator<T> accumulator = leaderState.accumulator();
         boolean isFirstAppend = accumulator.isEmpty();
-        final long offset = accumulator.append(epoch, records, requiredBaseOffset, isAtomic);
+        final long offset = accumulator.append(epoch, records, true);
```
Just curious, did we ever do non-atomic appends in the Raft layer before? I think the controller only did atomic appends (even though the controller itself would split up non-atomic batches)

jsancio replied:
The controller used to use both atomic and non-atomic append. That logic was changed to do the record splitting in the metadata module and only use atomic appends. This was required so that the in-memory snapshots generated by the controller were not split across batches.

If the records in an in-memory snapshot were split across batch boundaries then the controller was not able to revert to an in-memory snapshot.

Unrelated to the original motivation of this PR but I took the opportunity to remove API/methods that were not being used.

mumrah commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -817,27 +817,25 @@ public void run() throws Exception {
-                            records);
-                        offsetControl.handleScheduleAtomicAppend(nextEndOffset);
-                        return nextEndOffset;
+                        raftClient.scheduleFlush();
```
Any side effect if the batch already flushed by the time we get here? I suspect not.

jsancio replied:
Not possible. Flush it no accurate. I changed the name from `scheduleFlush` to `schedulePreparedAppend`.

`prepareAppend` accumulates batches in-memory and will never send them to the log layer until `schedulePreparedAppend` is called.

mumrah reviewed (commented):
@jsancio thank for the patch! I had a few questions after a brief review

If I understand this change, it looks like we are no longer tracking the next expected write offset and instead just appending to Raft to learn the last offset of the batch. 

1) Is it safe to schedule the write to Raft before the replay calls? I think it probably is. These two operations can tasks (writing to log and updating in-memory state) can probably be thought of as concurrent. As long as the in-memory state creates a snapshot with the correct offset, it should be fine. WDYT?

2) Do we need to track nextWriteOffset any more if we aren't using it? We the nextWriteOffset to set the broker epoch when registering brokers. I think this just needs to be monotonic relative to previous broker epochs. Maybe we can remove nextWriteOffset from OffsetControlManager and just track the "last written offset" instead?

3) In the previous code we tracked the next expected offset and passed it down to Raft each time we did an append. AIUI, this sort of acted like an optimistic concurrency control. E.g., "only do this operation if the next offset is X". We still have the controller epoch which serves a similar function, but I'm curious if there are side effects to removing the external offset tracking.

jsancio committed:
KAFKA-17304; Change the method names

jsancio committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-17304-kraft-explicit-flush

jsancio committed:
KAFKA-17304; Fix time to drain calculation

jsancio committed:
KAFKA-17304; Limit the maximum number of batches accumulated

jsancio commented:
> @jsancio thank for the patch! I had a few questions after a brief review
> 
> If I understand this change, it looks like we are no longer tracking the next expected write offset and instead just appending to Raft to learn the last offset of the batch.

I added a description to the PR and I think I address the motivation of this change there. Let me know if you have any other questions.

> 
> 1. Is it safe to schedule the write to Raft before the replay calls? I think it probably is. These two operations can tasks (writing to log and updating in-memory state) can probably be thought of as concurrent. As long as the in-memory state creates a snapshot with the correct offset, it should be fine. WDYT?

It is not safe to schedule write before replay but this change doesn't do that. At a high-level the new controller algorithm for write events is:
1. generate the cluster metadata records
2. stage them in the RaftClient so that the offsets are allocated (RaftClient#prepareAppend)
3. replay the cluster metadata records against the active controller state machine
4. tell the RaftClient client that is can write/append the staged batches to the log (RaftClient#schedulePreparedAppend)

> 2. Do we need to track nextWriteOffset any more if we aren't using it? We the nextWriteOffset to set the broker epoch when registering brokers. I think this just needs to be monotonic relative to previous broker epochs. Maybe we can remove nextWriteOffset from OffsetControlManager and just track the "last written offset" instead?

Isn't that the same thing? Isn't `"last written offset" = "next write offset" - 1`?

> 3. In the previous code we tracked the next expected offset and passed it down to Raft each time we did an append. AIUI, this sort of acted like an optimistic concurrency control. E.g., "only do this operation if the next offset is X". We still have the controller epoch which serves a similar function, but I'm curious if there are side effects to removing the external offset tracking.

Yes. The issue with that solution is that is doesn't allow for the KRaft leader to append control records to the log. Appending control records to the log would make it so that the controller known offset doesn't match the kraft leader's LEO. The KRaft leader needs to append control records to implement dynamically adding and removing controller/voters.

Take a look at the description of the PR. I have updated it since you last reviewed this PR. I tried to motivate the problem and explain the solution.

mumrah commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -817,27 +817,25 @@ public void run() throws Exception {
+                            long recordOffset = baseOffset + recordIndex;
                             try {
                                 replay(message.message(), Optional.empty(), recordOffset);
                             } catch (Throwable e) {
```
Do we need to do any cleanup to RaftClient if we hit a replay exception? Or is that handled by the resign code path?

jsancio replied:
Yeah. It is handle by the controller renounce code. The old code left it to the state machine (controller) to determine if a failure to write/append should result in a leader change. It is probably good to keep the same semantic unless we have a good reason to change it.

mumrah reviewed (approved):
Thanks for the updates and the PR description. It makes more sense to me now. I have one question inline, otherwise LGTM.

jsancio committed:
KAFKA-17304; Optimize kraft wake up
